File: Solutions/K-diff_Pairs_in_an_Array/K-diff_Pairs_in_an_Array.py
class Solution:
    def findPairs(self, nums, k):
        count = Counter(nums)
        if k > 0:
            res = sum([i + k in count for i in count])
        else:
            res = sum([count[i] > 1 for i in count])
        return res

#Fastest
class Solution:
    def findPairs(self, nums: List[int], k: int) -> int:
        if k != 0:
            s = set(nums)
            ans = 0
            for i in s:
                if i-k in s: ans += 1
            return ans
        else:
            s = set()
            used = set()
            ans = 0
            for i in nums:
                if i in s and i not in used: 
                    ans += 1
                    used.add(i)
                else: s.add(i)
            return ans
                

File: Solutions/K-diff_Pairs_in_an_Array/K-diff_Pairs_in_an_Array.cpp
class Solution
{

public:
    int findPairs(vector<int> &nums, int k)
    {

        unordered_map<int, int> mp;

        int count1 = 0;

        //putting all the values in hashmap
        for (int i = 0; i < nums.size(); i++)
        {
            mp[nums[i]]++;
        }

        //handling the case when k==0
        if (k == 0)
        {
            for (auto x : mp)
            {
                if (x.second > 1)
                    count1++;
            }
            return count1;
        }

        //we are iterating over hashmap because hashmap has unique values.
        for (auto x : mp)
        {
            if (mp.count(x.first - k))
                count1++;
        }

        return count1;
    }
};

//Second fastest
class Solution
{
public:
    int findPairs(vector<int> &nums, int k)
    {
        if (k < 0)
            return 0;
        if (k == 0)
        {
            std::unordered_map<int, int> myMap;

            for (auto e : nums)
                myMap[e]++;

            int max = 0;

            for (auto e : myMap)
            {
                if (e.second > 1)
                    max++;
            }

            return max;
        }

        std::unordered_set<int> mySet(nums.begin(), nums.end());

        int ans = 0;

        for (auto e : mySet)
        {
            if (mySet.find(e + k) != mySet.end())
                ans++;
        }
        return ans;
    }
};

//Fastest
class Solution
{
public:
    int findPairs(vector<int> &nums, int k)
    {
        sort(nums.begin(), nums.end());
        int i = 1;
        int j = 0;
        int res = 0;
        while (i < nums.size() && j < nums.size())
        {
            if (nums[i] - nums[j] == k)
            {
                res++;
                i++;
                j++;
                while (i < nums.size() && nums[i] == nums[i - 1])
                    i++;
            }
            else if (nums[i] - nums[j] > k)
            {
                j++;
                if (i - j == 0)
                    i++;
            }
            else
            {
                i++;
            }
        }
        return res;
    }
};